### PR TITLE
Add test for rustdoc-map generation when using sparse registries

### LINKED ISSUE #11402: Rustdoc-map generation fails when using sparse registries
### Problem

When using `-Zrustdoc-map` and `-Zsparse-registry` to document crates that are on an alternative sparse registry, cargo does not output the `--extern-html-root-url` values for any crates that are on that registry.

This is caused by the following equality check:
https://github.com/rust-lang/cargo/blob/65535890dbaa2b186633aedc11e550c3990fc5e9/src/cargo/core/compiler/rustdoc.rs#L135

When a dependency comes from a sparse registry, `sid.url()` has a scheme of `http`, but the `index_url` (which is based on the rustdoc-map configuration) has a scheme of `sparse+http`. Since the equality operation of the `Url` type considers the scheme components, the equality always fails.

<details><summary>Sample URL Objects</summary>
<p>

#### `sid.url()`

```rust
Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(1491), path: "/index/", query: None, fragment: None }
```

#### `index_url`

```rust
Url { scheme: "sparse+http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhost")), port: Some(1491), path: "/index/", query: None, fragment: None }
```

</p>
</details>

### Steps

1. Create an alternative registry that supports the sparse-registry interface.
2. Upload a library crate to that registry.
3. Create a new library crate that depends on the uploaded crate.
4. Add a `[doc.extern-map.registries]` key corresponding to the registry.
5. Run `cargo doc` with `-Zrustdoc-map`.

### Possible Solution(s)

The most straightforward fix is likely to compare if both URLs are sparse, and if so, compare the URLs with the `sparse+` qualifier removed from the scheme.

### Notes

_No response_

### Version

```text
cargo 1.67.0-nightly (eb5d35917 2022-11-17)
release: 1.67.0-nightly
commit-hash: eb5d35917b2395194593c9ca70c3778f60c1573b
commit-date: 2022-11-17
host: x86_64-unknown-linux-gnu
libgit2: 1.5.0 (sys:0.15.0 vendored)
libcurl: 7.86.0-DEV (sys:0.4.59+curl-7.86.0 vendored ssl:OpenSSL/1.1.1q)
os: Ubuntu 20.04 (focal) [64-bit]
```

### PR DESCRIPTION
**Pull request by calebfletcher**
Fixes #11402.

Adds an explicit check for sparse registries when generating the mappings for `-Zrustdoc-map`, so that the `sparse+` qualifier is removed before the comparison.

### CODE REVIEW

**Comment by calebfletcher** [comment]
Looks like #11387 could be related, and may fix the underlying issue with the SourceId

**Comment by weihanglo** [comment]
> Looks like #11387 could be related, and may fix the underlying issue with the SourceId

I am just about to say that 😆. It is still worth adding a test to verify the interaction between `-Zrustdoc-map` and `-Zsparse-registry, though.

**Comment by calebfletcher** [comment]
I've changed this pull request to add the requested test, note that the test currently fails but from my testing #11387 does correctly resolve the issue and the test passes with those changes applied. As such, I have marked this PR as a draft and once #11387 lands I will mark it as ready for review again.

**Comment by calebfletcher** [comment]
Running the test locally on master after that PR got merged seems to work, so not sure why the CI failed. Have rebased to see if that clears it.

**Comment by calebfletcher** [comment]
@weihanglo looks like all the tests are happy now, so the change in #11387 definitely did fix this issue as well

File: tests/testsuite/rustdoc_extern_html.rs [hunk 1]
```diff
@@ -342,3 +342,85 @@ fn rebuilds_when_changing() {
...
+            [..]bar=https://example.com/bar/1.0.0/[..]grimm=https://docs.rs/grimm/1.0.0/[..]",
+        )
+        .run();
+    let queen = p.read_file("target/doc/foo/fn.queen.html");
+    assert!(queen.contains(r#"href="https://example.com/bar/1.0.0/bar/struct.Queen.html""#));
+    // The king example fails to link. Rustdoc seems to want the origin crate
+    // name (baz) for re-exports. There are many issues in the issue tracker
+    // for rustdoc re-exports, so I'm not sure, but I think this is maybe a
+    // rustdoc issue. Alternatively, Cargo could provide mappings for all
+    // transitive dependencies to fix this.
```
Comment (weihanglo) [inline tests/testsuite/rustdoc_extern_html.rs:420]: It might be, but let's move on first :)

**Review by weihanglo — APPROVED** [review]
Thank you for adding such a test!

File: tests/testsuite/rustdoc_extern_html.rs (hunk 1 above)
Comment (calebfletcher) [reply tests/testsuite/rustdoc_extern_html.rs:420]: Ah my bad, that was a comment that is in the test that I adapted this from (you can see it in the original test here:
https://github.com/rust-lang/cargo/blob/f54cd332e2d8fa6359ec226f150284256d0e540d/tests/testsuite/rustdoc_extern_html.rs#L262)

**Pull request merged**